### PR TITLE
Check task dependencies before processing the task

### LINKED ISSUE #756: Task won't start with volumesFrom defined, agent doesn't say why
I have a basic task defined with 2 containers - an nginx container `site` linked to a php-fpm  container `site-php`. The php-fpm container needs to access the website files on the nginx container so I'm trying to use `volumesFrom` to do this. 

Task definition is:
```json
{
    "family": "website",
    "networkMode": "bridge",
    "containerDefinitions": [{
        "name": "site",
        "memoryReservation": 500,
        "portMappings": [{
            "containerPort": 80,
            "hostPort": 80
        }, {
            "containerPort": 443,
            "hostPort": 443
        }],
        "essential": true,
        "links": ["site-php:site-php"],
        "image": "372423516982.dkr.ecr.eu-west-2.amazonaws.com/site/www:latest"
    }, {
        "name": "site-php",
        "volumesFrom": [{
            "sourceContainer": "site"
        }],
        "memoryReservation": 1000,
        "image": "372423516982.dkr.ecr.eu-west-2.amazonaws.com/site/www-php:latest"
    }]
}
```

The Dockerfile for `site` has the line `VOLUME /var/www/html` which I'm expecting the `site-php` container to mount. The exact same task definition with the `volumesFrom` section removed from `site-php` starts up fine (although the website obviously doesn't work I can still get a response from nginx) and the docker-compose equivalent of this run locally works too. 

When I run the task with `volumesFrom` defined this way it stops immediately, the agent logs the following but no containers are started and there're no other logs available to work out what the problem might be. The AWS interface just says `Stopped reason: Essential container in task exited` with no further information.

```
2017-04-06T11:45:20Z [CRITICAL] Task in a bad state; it's not steadystate but no containers want to transition module="TaskEngine" task="site-only:4 arn:aws:ecs:eu-west-2:372423516982:task/c9ce3c1d-97b2-4656-96da-a576059403d1, Status: (NONE->RUNNING) Containers: [site (NONE->RUNNING),site-php (NONE->RUNNING),]"
2017-04-06T11:45:20Z [CRITICAL] Moving task to stopped due to bad state module="TaskEngine" task="site-only:4 arn:aws:ecs:eu-west-2:372423516982:task/c9ce3c1d-97b2-4656-96da-a576059403d1, Status: (NONE->RUNNING) Containers: [site (NONE->RUNNING),site-php (NONE->RUNNING),]"
```

Is there something obvious wrong with this task definition? If not, how can I get better error messages to work out what the problem is?

### PR DESCRIPTION
**Pull request by richardpen**
### Summary

Fix #756 Check the task dependencies before processing the task and provide a more clear log message.

### Implementation details

Check the dependencies before processing the task in task engine.

### Testing

- [ ] Builds on Linux (`make release`)
- [ ] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [ ] Unit tests on Linux (`make test`) pass
- [ ] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [ ] Integration tests on Linux (`make run-integ-tests`) pass
- [ ] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [ ] Functional tests on Linux (`make run-functional-tests`) pass
- [ ] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: 
yes
### Licensing

This contribution is under the terms of the Apache 2.0 License: 
yes

### CODE REVIEW

File: agent/engine/docker_task_engine_test.go [hunk 1]
```diff
@@ -1146,3 +1146,27 @@ func TestEngineDisableConcurrentPull(t *testing.T) {
 	assert.False(t, dockerTaskEngine.enableConcurrentPull,
 		"Task engine should not be able to perform concurrent pulling for version < 1.11.1")
 }
+
+// TestTaskWithCircularDependency tests the task with containers of which the
```
Comment (aaithal) [inline agent/engine/docker_task_engine_test.go]: Can you also add a unit test in `dependencygraph` for `ValidDependencies()` to test a task with circular dependencies?

File: agent/engine/docker_task_engine.go [hunk 2]
```diff
@@ -447,8 +448,19 @@ func (engine *DockerTaskEngine) AddTask(task *api.Task) error {
...
+
 		engine.state.AddTask(task)
-		engine.startTask(task)
+		if dependencygraph.ValidDependencies(task) {
+			engine.startTask(task)
+		} else {
+			seelog.Errorf("Task can not move forward due to the dependencies of containers cannot be resolved, task:%s", task.String())
+			task.SetKnownStatus(api.TaskStopped)
+			task.SetDesiredStatus(api.TaskStopped)
+			engine.emitTaskEvent(task, "TaskTransitionError: Can not resolve the dependencies of containers in the task definition")
```
Comment (aaithal) [inline agent/engine/docker_task_engine.go]: can you use the `impossibleTransitionError` type here instead?

File: agent/engine/docker_task_engine.go [hunk 3]
```diff
@@ -447,8 +448,19 @@ func (engine *DockerTaskEngine) AddTask(task *api.Task) error {
...
 	if !exists {
+		// This will update the container desired status
+		task.UpdateDesiredStatus()
+
 		engine.state.AddTask(task)
-		engine.startTask(task)
+		if dependencygraph.ValidDependencies(task) {
+			engine.startTask(task)
+		} else {
+			seelog.Errorf("Task can not move forward due to the dependencies of containers cannot be resolved, task:%s", task.String())
```
Comment (aaithal) [inline agent/engine/docker_task_engine.go]: We should either use `can not` or `cannot`. Let's not use both. I'd vote for `cannot`. Also, can you leave a space after `task: `?

File: agent/engine/docker_task_engine.go [hunk 4]
```diff
@@ -447,8 +448,19 @@ func (engine *DockerTaskEngine) AddTask(task *api.Task) error {
 
 	existingTask, exists := engine.state.TaskByArn(task.Arn)
 	if !exists {
+		// This will update the container desired status
+		task.UpdateDesiredStatus()
+
 		engine.state.AddTask(task)
-		engine.startTask(task)
+		if dependencygraph.ValidDependencies(task) {
+			engine.startTask(task)
```
Comment (aaithal) [inline agent/engine/docker_task_engine.go]: you can also just `return nil` here and get rid of `else` in line 464.

Comment (jhaynes) [inline agent/engine/docker_task_engine.go]: @aaithal Can you explain that? There is no codepath where we could call `AddTask` on an existing task?

File: agent/engine/docker_task_engine.go (hunk 2 above)
Comment (jhaynes) [inline agent/engine/docker_task_engine.go]: Cannot vs. Can not. Cannot is "more correct"

File: agent/engine/docker_task_engine_test.go (hunk 1 above)
Comment (richardpen) [inline agent/engine/docker_task_engine_test.go]: Weren't this have already tested it: https://github.com/aws/amazon-ecs-agent/blob/master/agent/engine/dependencygraph/graph_test.go#L48

File: agent/engine/docker_task_engine.go (hunk 4 above)
Comment (richardpen) [inline agent/engine/docker_task_engine.go]: I think @aaithal means get rid of the `else` word but keeps the statement inside the else, see my updates.

File: agent/engine/docker_task_engine_test.go (hunk 1 above)
Comment (jahkeup) [inline agent/engine/docker_task_engine_test.go]: Indeed, this sort of cyclical dependency is being tested [here](https://github.com/aws/amazon-ecs-agent/blob/467c3d772d55e02ad618e4901afed1da1173496e/agent/engine/dependencygraph/graph_test.go#L88-L98).

**Comment by vsiddharth** [comment]
@richardpen these changes look good to me

File: agent/engine/docker_task_engine.go [hunk 5]
```diff
@@ -446,12 +447,24 @@ func (engine *DockerTaskEngine) AddTask(task *api.Task) error {
...
-		engine.startTask(task)
-	} else {
-		engine.updateTask(existingTask, task)
+		if dependencygraph.ValidDependencies(task) {
+			engine.startTask(task)
+		} else {
+			seelog.Errorf("Task cannot move forward due to the dependencies of containers cannot be resolved, task: %s", task.String())
+			task.SetKnownStatus(api.TaskStopped)
+			task.SetDesiredStatus(api.TaskStopped)
+			engine.emitTaskEvent(task, "impossibleTransitionError: Cannot resolve the dependencies of containers in the task definition")
```
Comment (aaithal) [inline agent/engine/docker_task_engine.go]: Apologies, I should have made this clearer in my earlier comment. Please use the `impossibleTransitionError ` type instead: https://github.com/aws/amazon-ecs-agent/blob/f88e52ec2d20fd65b2ccffc0019cc1aa203e487d/agent/engine/errors.go#L35 instead.

File: agent/engine/errors.go [hunk 6]
```diff
@@ -112,6 +112,21 @@ func (CannotGetDockerClientError) ErrorName() string {
...
+	taskArn string
+}
+
+func (err TaskCircularDependencyError) Error() string {
+	return "Task has circular dependencies, cannot be processed" + err.taskArn
+}
+
+// ErrorName is the name of the error
+func (err TaskCircularDependencyError) ErrorName() string {
+	return "TAskCircularDependencyError"
```
Comment (aaithal) [inline agent/engine/errors.go]: typo: `"Task"`

File: agent/engine/errors.go [hunk 7]
```diff
@@ -112,6 +112,21 @@ func (CannotGetDockerClientError) ErrorName() string {
 	return "CannotGetDockerclientError"
 }
 
+// TaskCircularDependencyError is the error for task that dependencies can't
+// be resolved
+type TaskCircularDependencyError struct {
```
Comment (aaithal) [inline agent/engine/errors.go]: I wonder if `TaskDependencyError` is a better type name. Because, the  `ValidDependencies` method might fail for other reasons as well.

File: agent/engine/docker_task_engine.go [hunk 8]
```diff
@@ -446,12 +447,25 @@ func (engine *DockerTaskEngine) AddTask(task *api.Task) error {
...
+		task.UpdateDesiredStatus()
+
 		engine.state.AddTask(task)
-		engine.startTask(task)
-	} else {
-		engine.updateTask(existingTask, task)
+		if dependencygraph.ValidDependencies(task) {
+			engine.startTask(task)
+		} else {
+			seelog.Errorf("Task cannot move forward due to the dependencies of containers cannot be resolved, task: %s", task.String())
```
Comment (samuelkarp) [inline agent/engine/docker_task_engine.go]: Can you rephrase this?  "Unable to progress task with circular dependencies"

File: agent/engine/errors.go [hunk 9]
```diff
@@ -112,6 +112,21 @@ func (CannotGetDockerClientError) ErrorName() string {
...
 }
 
+// TaskDependencyError is the error for task that dependencies can't
+// be resolved
+type TaskDependencyError struct {
+	taskArn string
+}
+
+func (err TaskDependencyError) Error() string {
+	return "Task dependencies cannot be resolved" + err.taskArn
```
Comment (samuelkarp) [inline agent/engine/errors.go]: You need a space at the end here or you'll end up with "resolvedarn:"
